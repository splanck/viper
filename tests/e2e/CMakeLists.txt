# SPDX-License-Identifier: MIT
# File: tests/e2e/CMakeLists.txt
# Purpose: End-to-end CLI and VM integration tests.

function(viper_add_e2e_vm_examples_smoke)
  if(TEST vm_examples)
    return()
  endif()

  set(_VIPER_E2E_DIR ${CMAKE_CURRENT_FUNCTION_LIST_DIR})
  viper_add_ctest(vm_examples
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DSRC_DIR=${CMAKE_SOURCE_DIR}
    -P ${_VIPER_E2E_DIR}/test_vm_examples.cmake)
  set_tests_properties(vm_examples PROPERTIES LABELS VM)
endfunction()

function(viper_add_e2e_vm_suite)
  if(TEST vm_strings_example)
    return()
  endif()

  set(_VIPER_E2E_DIR ${CMAKE_CURRENT_FUNCTION_LIST_DIR})

  if(NOT TARGET float_out)
    viper_add_test(float_out ${_VIPER_E2E_DIR}/support/FloatOut.cpp NO_CTEST)
  endif()

  viper_add_ctest(vm_strings_example
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DSRC_DIR=${CMAKE_SOURCE_DIR}
    -P ${_VIPER_E2E_DIR}/test_vm_strings.cmake)
  set_tests_properties(vm_strings_example PROPERTIES LABELS VM)

  viper_add_ctest(vm_step_limit
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DSRC_DIR=${CMAKE_SOURCE_DIR}
    -P ${_VIPER_E2E_DIR}/test_vm_step_limit.cmake)
  set_tests_properties(vm_step_limit PROPERTIES LABELS VM)

  viper_add_ctest(vm_block_params
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DSRC_DIR=${CMAKE_SOURCE_DIR}
    -P ${_VIPER_E2E_DIR}/test_vm_block_params.cmake)
  set_tests_properties(vm_block_params PROPERTIES LABELS VM)

  viper_add_ctest(vm_math_core
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DSRC_DIR=${CMAKE_SOURCE_DIR}
    -P ${_VIPER_E2E_DIR}/test_vm_math_core.cmake)
  set_tests_properties(vm_math_core PROPERTIES LABELS VM)

  viper_add_ctest(vm_math_trigpow
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DSRC_DIR=${CMAKE_SOURCE_DIR}
    -P ${_VIPER_E2E_DIR}/test_vm_math_trigpow.cmake)
  set_tests_properties(vm_math_trigpow PROPERTIES LABELS VM)

  viper_add_ctest(vm_rec_fact
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DSRC_DIR=${CMAKE_SOURCE_DIR}
    -P ${_VIPER_E2E_DIR}/test_vm_rec_fact.cmake)
  set_tests_properties(vm_rec_fact PROPERTIES LABELS VM)

  viper_add_ctest(vm_trace_src
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/examples/basic/trace_src.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/e2e/TraceSrcTests.bas.out
    -P ${_VIPER_E2E_DIR}/test_trace_src.cmake)
  set_tests_properties(vm_trace_src PROPERTIES LABELS VM)

  viper_add_ctest(vm_break_src_exact
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DSRC_FILE=tests/e2e/BreakSrcExact.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/golden/break_src_exact.out
    -DLINE=1
    -DROOT=${CMAKE_SOURCE_DIR}
    -P ${_VIPER_E2E_DIR}/test_break_src_exact.cmake)
  set_tests_properties(vm_break_src_exact PROPERTIES LABELS VM DISABLED TRUE)

  viper_add_ctest(vm_break_src_coalesce
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DSRC_FILE=tests/e2e/BreakSrcCoalesce.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/golden/break_src_coalesce.out
    -DLINE=1
    -DROOT=${CMAKE_SOURCE_DIR}
    -P ${_VIPER_E2E_DIR}/test_break_src_exact.cmake)
  set_tests_properties(vm_break_src_coalesce PROPERTIES LABELS VM)

  viper_add_ctest(front_basic_example
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DSRC_DIR=${CMAKE_SOURCE_DIR}
    -P ${_VIPER_E2E_DIR}/test_front_basic.cmake)

  viper_add_ctest(basic_math_phase1
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/tests/e2e/basic_math_phase1.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/e2e/basic_math_phase1.bas.out
    -P ${_VIPER_E2E_DIR}/test_basic_math_phase1.cmake)

  viper_add_ctest(basic_math_phase2
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/tests/e2e/basic_math_phase2.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/e2e/basic_math_phase2.bas.out
    -P ${_VIPER_E2E_DIR}/test_basic_math_phase1.cmake)

  viper_add_ctest(basic_abs_mixed
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/tests/e2e/basic_abs_mixed.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/e2e/basic_abs_mixed.bas.out
    -P ${_VIPER_E2E_DIR}/test_basic_math_phase1.cmake)

  viper_add_ctest(basic_numerics_addmul
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/tests/e2e/basic/numerics_addmul.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/e2e/basic/numerics_addmul.bas.out
    -DEXPECT_SUBSTR=Overflow
    -P ${_VIPER_E2E_DIR}/test_basic_numeric_overflow.cmake)
  set_tests_properties(basic_numerics_addmul PROPERTIES LABELS VM)

  viper_add_ctest(basic_numerics_pow
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/tests/e2e/basic/numerics_pow.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/e2e/basic/numerics_pow.bas.out
    -DEXPECT_SUBSTR=DomainError
    -P ${_VIPER_E2E_DIR}/test_basic_numeric_overflow.cmake)
  set_tests_properties(basic_numerics_pow PROPERTIES LABELS VM)

  viper_add_ctest(basic_numerics_conversions
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/tests/e2e/basic/numerics_conversions.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/e2e/basic/numerics_conversions.bas.out
    -DEXPECT_SUBSTR=InvalidCast
    -P ${_VIPER_E2E_DIR}/test_basic_numeric_overflow.cmake)
  set_tests_properties(basic_numerics_conversions PROPERTIES LABELS VM)

  viper_add_ctest(basic_numerics_val
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/tests/e2e/basic/numerics_val.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/e2e/basic/numerics_val.bas.out
    -P ${_VIPER_E2E_DIR}/test_basic_math_phase1.cmake)
  set_tests_properties(basic_numerics_val PROPERTIES LABELS VM)

  viper_add_ctest(basic_array_sum
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/tests/e2e/basic_array_sum.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/e2e/basic_array_sum.bas.out
    -P ${_VIPER_E2E_DIR}/test_basic_array_sum.cmake)
  set_tests_properties(basic_array_sum PROPERTIES LABELS VM)

  viper_add_ctest(basic_do_exit
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/tests/e2e/do_exit.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/e2e/do_exit.bas.out
    -P ${_VIPER_E2E_DIR}/test_do_exit.cmake)
  set_tests_properties(basic_do_exit PROPERTIES LABELS VM)

  viper_add_ctest(basic_for_step_neg
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/tests/e2e/for_step_neg.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/e2e/for_step_neg.bas.out
    -P ${_VIPER_E2E_DIR}/test_do_exit.cmake)
  set_tests_properties(basic_for_step_neg PROPERTIES LABELS VM)

  viper_add_ctest(basic_array_oob
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/tests/e2e/basic_array_oob.bas
    "-DEXPECT_SUBSTR=rt_arr_i32: index 3 out of bounds (len=3)"
    -P ${_VIPER_E2E_DIR}/test_basic_array_oob.cmake)
  set_tests_properties(basic_array_oob PROPERTIES LABELS VM)

  viper_add_ctest(basic_factorial
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/tests/e2e/factorial.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/golden/e2e/factorial.out
    -P ${_VIPER_E2E_DIR}/test_basic_math_phase1.cmake)
  set_tests_properties(basic_factorial PROPERTIES WILL_FAIL TRUE)

  viper_add_ctest(basic_fibonacci
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/tests/e2e/fibonacci.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/golden/e2e/fibonacci.out
    -P ${_VIPER_E2E_DIR}/test_basic_math_phase1.cmake)
  set_tests_properties(basic_fibonacci PROPERTIES WILL_FAIL TRUE)

  viper_add_ctest(basic_random_repro
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DSRC_DIR=${CMAKE_SOURCE_DIR}
    -P ${_VIPER_E2E_DIR}/test_basic_random_repro.cmake)

  viper_add_ctest(basic_numerics_all
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/tests/e2e/basic/numerics_all.bas
    -DGOLDEN=${CMAKE_SOURCE_DIR}/tests/e2e/basic/numerics_all.bas.out
    -P ${_VIPER_E2E_DIR}/test_basic_math_phase1.cmake)

  viper_add_ctest(monte_carlo_pi
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DFLOAT_OUT=$<TARGET_FILE:float_out>
    -DSRC_DIR=${CMAKE_SOURCE_DIR}
    -P ${_VIPER_E2E_DIR}/test_monte_carlo_pi.cmake)

  viper_add_ctest(random_walk
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DSRC_DIR=${CMAKE_SOURCE_DIR}
    -P ${_VIPER_E2E_DIR}/test_random_walk.cmake)

  viper_add_ctest(il_opt_equiv
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DIL_FILE=${CMAKE_SOURCE_DIR}/tests/golden/il_opt/e2e.in.il
    -P ${_VIPER_E2E_DIR}/test_il_opt_equiv.cmake)

  viper_add_ctest(constfold_equiv
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DIL_FILE=${CMAKE_SOURCE_DIR}/tests/il/e2e/math_constfold.il
    -P ${_VIPER_E2E_DIR}/test_constfold_equiv.cmake)

  viper_add_ctest(mem2reg_basic
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DBAS_FILE=${CMAKE_SOURCE_DIR}/tests/data/mem2reg.bas
    -P ${_VIPER_E2E_DIR}/test_mem2reg_basic.cmake)

  viper_add_ctest(ilc_mem2reg_stats
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DIL_FILE=${CMAKE_SOURCE_DIR}/tests/golden/il_opt/mem2reg_simple.in.il
    -P ${_VIPER_E2E_DIR}/test_mem2reg_stats.cmake)

  viper_add_ctest(ilc_mem2reg_stats_multi
    ${CMAKE_COMMAND}
    -DILC=$<TARGET_FILE:ilc>
    -DIL_FILE=${CMAKE_SOURCE_DIR}/tests/golden/il_opt/mem2reg_multi.in.il
    -P ${_VIPER_E2E_DIR}/test_mem2reg_stats_multi.cmake)
endfunction()

function(viper_add_e2e_codegen_tests)
  if(DEFINED VIPER_E2E_CODEGEN_DONE)
    return()
  endif()

  set(_VIPER_E2E_DIR ${CMAKE_CURRENT_FUNCTION_LIST_DIR})
  set(CODEGEN_E2E ${_VIPER_E2E_DIR}/test_codegen.cmake)

  if(IL_ENABLE_X64_ASM_SYNTAX_CHECK AND CLANG_FOUND)
    viper_add_ctest(codegen_syntax_only
      ${CMAKE_COMMAND}
      "-DASM_FLAGS=${IL_X64_ASM_FLAGS}"
      "-DLD_FLAGS=${IL_X64_LD_FLAGS}"
      -DMODE=syntax
      -P ${CODEGEN_E2E})
    set_tests_properties(codegen_syntax_only PROPERTIES LABELS CodegenSyntaxOnly)
  else()
    message(STATUS "Skipping codegen_* tests (CLANG not found or flags OFF)")
  endif()

  if(IL_ENABLE_X64_ASM_ASSEMBLE_LINK AND CLANG_FOUND)
    viper_add_ctest(codegen_assemble_link
      ${CMAKE_COMMAND}
      "-DASM_FLAGS=${IL_X64_ASM_FLAGS}"
      "-DLD_FLAGS=${IL_X64_LD_FLAGS}"
      -DMODE=assemble_link
      -P ${CODEGEN_E2E})
    set_tests_properties(codegen_assemble_link PROPERTIES LABELS CodegenAssembleLink)
  else()
    message(STATUS "Skipping codegen_* tests (CLANG not found or flags OFF)")
  endif()

  if(IL_ENABLE_X64_NATIVE_RUN AND CLANG_FOUND)
    viper_add_ctest(codegen_native_run
      ${CMAKE_COMMAND}
      "-DASM_FLAGS=${IL_X64_ASM_FLAGS}"
      "-DLD_FLAGS=${IL_X64_LD_FLAGS}"
      -DMODE=run
      -P ${CODEGEN_E2E})
    set_tests_properties(codegen_native_run PROPERTIES LABELS NativeX64Run)
  else()
    message(STATUS "Skipping codegen_* tests (CLANG not found or flags OFF)")
  endif()

  set(VIPER_E2E_CODEGEN_DONE TRUE PARENT_SCOPE)
endfunction()
