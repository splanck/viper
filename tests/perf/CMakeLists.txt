# SPDX-License-Identifier: MIT
# File: tests/perf/CMakeLists.txt
# Purpose: Performance-oriented regression checks.

viper_add_test(
  test_perf_vm_switch
  ${CMAKE_CURRENT_LIST_DIR}/vm_switch_bench.cpp
  TEST_NAME perf_vm_switch_bench)
target_link_libraries(test_perf_vm_switch PRIVATE il_build il_vm viper_support)
