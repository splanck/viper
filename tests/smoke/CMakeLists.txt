# SPDX-License-Identifier: MIT
# File: tests/smoke/CMakeLists.txt
# Purpose: CLI smoke tests for terminal BASIC programs.

add_test(NAME smoke_term_basic
         COMMAND $<TARGET_FILE:ilc> front basic -run ${CMAKE_CURRENT_SOURCE_DIR}/term_basic.bas)
