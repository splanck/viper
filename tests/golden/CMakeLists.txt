# SPDX-License-Identifier: MIT
# File: tests/golden/CMakeLists.txt
# Purpose: Golden tests for BASIC diagnostics and IL optimizations.

function(viper_add_golden_suite_tests)
  if(TEST basic_error_unknown_var)
    return()
  endif()

  set(_VIPER_GOLDEN_DIR ${CMAKE_CURRENT_FUNCTION_LIST_DIR})
  set(BASIC_ILC $<TARGET_FILE:ilc>)

  file(READ ${_VIPER_GOLDEN_DIR}/basic_errors/unknown_var.stderr EXPECT_UNKNOWN_VAR)
  string(STRIP "${EXPECT_UNKNOWN_VAR}" EXPECT_UNKNOWN_VAR)
  viper_add_ctest(basic_error_unknown_var
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DBAS_FILE=${_VIPER_GOLDEN_DIR}/basic_errors/unknown_var.bas
    "-DEXPECT=${EXPECT_UNKNOWN_VAR}"
    -P ${_VIPER_GOLDEN_DIR}/basic_errors/check_error.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/basic_errors/mismatched_next.stderr EXPECT_MISMATCHED_NEXT)
  string(STRIP "${EXPECT_MISMATCHED_NEXT}" EXPECT_MISMATCHED_NEXT)
  viper_add_ctest(basic_error_mismatched_next
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DBAS_FILE=${_VIPER_GOLDEN_DIR}/basic_errors/mismatched_next.bas
    "-DEXPECT=${EXPECT_MISMATCHED_NEXT}"
    -P ${_VIPER_GOLDEN_DIR}/basic_errors/check_error.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/basic_errors/loop_var_assignment.stderr EXPECT_LOOP_VAR_ASSIGNMENT)
  string(STRIP "${EXPECT_LOOP_VAR_ASSIGNMENT}" EXPECT_LOOP_VAR_ASSIGNMENT)
  viper_add_ctest(basic_error_loop_var_assignment
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DBAS_FILE=${_VIPER_GOLDEN_DIR}/basic_errors/loop_var_assignment.bas
    "-DEXPECT=${EXPECT_LOOP_VAR_ASSIGNMENT}"
    -P ${_VIPER_GOLDEN_DIR}/basic_errors/check_error.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/basic_errors/loop_var_input.stderr EXPECT_LOOP_VAR_INPUT)
  string(STRIP "${EXPECT_LOOP_VAR_INPUT}" EXPECT_LOOP_VAR_INPUT)
  viper_add_ctest(basic_error_loop_var_input
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DBAS_FILE=${_VIPER_GOLDEN_DIR}/basic_errors/loop_var_input.bas
    "-DEXPECT=${EXPECT_LOOP_VAR_INPUT}"
    -P ${_VIPER_GOLDEN_DIR}/basic_errors/check_error.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/basic_errors/bad_goto.stderr EXPECT_BAD_GOTO)
  string(STRIP "${EXPECT_BAD_GOTO}" EXPECT_BAD_GOTO)
  viper_add_ctest(basic_error_bad_goto
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DBAS_FILE=${_VIPER_GOLDEN_DIR}/basic_errors/bad_goto.bas
    "-DEXPECT=${EXPECT_BAD_GOTO}"
    -P ${_VIPER_GOLDEN_DIR}/basic_errors/check_error.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/basic_errors/string_arith.stderr EXPECT_STRING_ARITH)
  string(STRIP "${EXPECT_STRING_ARITH}" EXPECT_STRING_ARITH)
  viper_add_ctest(basic_error_string_arith
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DBAS_FILE=${_VIPER_GOLDEN_DIR}/basic_errors/string_arith.bas
    "-DEXPECT=${EXPECT_STRING_ARITH}"
    -P ${_VIPER_GOLDEN_DIR}/basic_errors/check_error.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/basic_errors/string_cmp_order.stderr EXPECT_STRING_CMP_ORDER)
  string(STRIP "${EXPECT_STRING_CMP_ORDER}" EXPECT_STRING_CMP_ORDER)
  viper_add_ctest(basic_error_string_cmp_order
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DBAS_FILE=${_VIPER_GOLDEN_DIR}/basic_errors/string_cmp_order.bas
    "-DEXPECT=${EXPECT_STRING_CMP_ORDER}"
    -P ${_VIPER_GOLDEN_DIR}/basic_errors/check_error.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/basic_errors/const_div_zero.stderr EXPECT_CONST_DIV_ZERO)
  string(STRIP "${EXPECT_CONST_DIV_ZERO}" EXPECT_CONST_DIV_ZERO)
  viper_add_ctest(basic_error_const_div_zero
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DBAS_FILE=${_VIPER_GOLDEN_DIR}/basic_errors/const_div_zero.bas
    "-DEXPECT=${EXPECT_CONST_DIV_ZERO}"
    -P ${_VIPER_GOLDEN_DIR}/basic_errors/check_error.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/basic_errors/const_divmod_zero.stderr EXPECT_CONST_DIVMOD_ZERO)
  string(STRIP "${EXPECT_CONST_DIVMOD_ZERO}" EXPECT_CONST_DIVMOD_ZERO)
  viper_add_ctest(basic_error_const_divmod_zero
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DBAS_FILE=${_VIPER_GOLDEN_DIR}/basic_errors/const_divmod_zero.bas
    "-DEXPECT=${EXPECT_CONST_DIVMOD_ZERO}"
    -P ${_VIPER_GOLDEN_DIR}/basic_errors/check_error.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/basic_errors/string_divmod.stderr EXPECT_STRING_DIVMOD)
  string(STRIP "${EXPECT_STRING_DIVMOD}" EXPECT_STRING_DIVMOD)
  viper_add_ctest(basic_error_string_divmod
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DBAS_FILE=${_VIPER_GOLDEN_DIR}/basic_errors/string_divmod.bas
    "-DEXPECT=${EXPECT_STRING_DIVMOD}"
    -P ${_VIPER_GOLDEN_DIR}/basic_errors/check_error.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/basic_errors/bad_string_intrinsics.stderr EXPECT_BAD_STRING_INTRINSICS)
  string(STRIP "${EXPECT_BAD_STRING_INTRINSICS}" EXPECT_BAD_STRING_INTRINSICS)
  viper_add_ctest(basic_error_bad_string_intrinsics
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DBAS_FILE=${_VIPER_GOLDEN_DIR}/basic_errors/bad_string_intrinsics.bas
    "-DEXPECT=${EXPECT_BAD_STRING_INTRINSICS}"
    -P ${_VIPER_GOLDEN_DIR}/basic_errors/check_error.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/basic_errors/float_to_int.stderr EXPECT_FLOAT_TO_INT)
  string(STRIP "${EXPECT_FLOAT_TO_INT}" EXPECT_FLOAT_TO_INT)
  viper_add_ctest(basic_error_float_to_int
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DBAS_FILE=${_VIPER_GOLDEN_DIR}/basic_errors/float_to_int.bas
    "-DEXPECT=${EXPECT_FLOAT_TO_INT}"
    -P ${_VIPER_GOLDEN_DIR}/basic_errors/check_error.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/basic_errors/float_in_string.stderr EXPECT_FLOAT_IN_STRING)
  string(STRIP "${EXPECT_FLOAT_IN_STRING}" EXPECT_FLOAT_IN_STRING)
  viper_add_ctest(basic_error_float_in_string
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DBAS_FILE=${_VIPER_GOLDEN_DIR}/basic_errors/float_in_string.bas
    "-DEXPECT=${EXPECT_FLOAT_IN_STRING}"
    -P ${_VIPER_GOLDEN_DIR}/basic_errors/check_error.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/basic_errors/bad_conversions.stderr EXPECT_BAD_CONVERSIONS)
  string(STRIP "${EXPECT_BAD_CONVERSIONS}" EXPECT_BAD_CONVERSIONS)
  viper_add_ctest(basic_error_bad_conversions
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DBAS_FILE=${_VIPER_GOLDEN_DIR}/basic_errors/bad_conversions.bas
    "-DEXPECT=${EXPECT_BAD_CONVERSIONS}"
    -P ${_VIPER_GOLDEN_DIR}/basic_errors/check_error.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/basic_errors/bad_math_funcs.stderr EXPECT_BAD_MATH_FUNCS)
  string(STRIP "${EXPECT_BAD_MATH_FUNCS}" EXPECT_BAD_MATH_FUNCS)
  viper_add_ctest(basic_error_bad_math_funcs
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DBAS_FILE=${_VIPER_GOLDEN_DIR}/basic_errors/bad_math_funcs.bas
    "-DEXPECT=${EXPECT_BAD_MATH_FUNCS}"
    -P ${_VIPER_GOLDEN_DIR}/basic_errors/check_error.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/vm_trap_loc.stderr EXPECT_VM_TRAP_LOC)
  string(STRIP "${EXPECT_VM_TRAP_LOC}" EXPECT_VM_TRAP_LOC)
  viper_add_ctest(vm_trap_loc
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DIL_FILE=${_VIPER_GOLDEN_DIR}/vm_trap_loc.il
    "-DEXPECT=${EXPECT_VM_TRAP_LOC}"
    -P ${_VIPER_GOLDEN_DIR}/check_vm_trap_loc.cmake)

  viper_add_ctest(il_opt_const_if
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DIL_FILE=${_VIPER_GOLDEN_DIR}/il_opt/const_if.in.il
    -DGOLDEN=${_VIPER_GOLDEN_DIR}/il_opt/const_if.opt.il
    -P ${_VIPER_GOLDEN_DIR}/il_opt/check_opt.cmake)

  viper_add_ctest(il_opt_arith_id
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DIL_FILE=${_VIPER_GOLDEN_DIR}/il_opt/arith_id.in.il
    -DGOLDEN=${_VIPER_GOLDEN_DIR}/il_opt/arith_id.opt.il
    -P ${_VIPER_GOLDEN_DIR}/il_opt/check_opt.cmake)

  viper_add_ctest(il_opt_math_constfold
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DIL_FILE=${_VIPER_GOLDEN_DIR}/il_opt/math_constfold.in.il
    -DGOLDEN=${_VIPER_GOLDEN_DIR}/il_opt/math_constfold.opt.il
    -DPASSES=constfold
    -P ${_VIPER_GOLDEN_DIR}/il_opt/check_opt.cmake)

  viper_add_ctest(il_opt_cbr_same_target
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DIL_FILE=${_VIPER_GOLDEN_DIR}/il_opt/cbr_same_target.in.il
    -DGOLDEN=${_VIPER_GOLDEN_DIR}/il_opt/cbr_same_target.golden
    -DPASSES=peephole
    -P ${_VIPER_GOLDEN_DIR}/il_opt/check_opt.cmake)
  set_tests_properties(il_opt_cbr_same_target PROPERTIES LABELS Smoke)

  viper_add_ctest(il_opt_mem2reg_simple
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DIL_FILE=${_VIPER_GOLDEN_DIR}/il_opt/mem2reg_simple.in.il
    -DGOLDEN=${_VIPER_GOLDEN_DIR}/il_opt/mem2reg_simple.opt.il
    -DPASSES=mem2reg
    -P ${_VIPER_GOLDEN_DIR}/il_opt/check_opt.cmake)

  viper_add_ctest(il_opt_mem2reg_diamond
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DIL_FILE=${_VIPER_GOLDEN_DIR}/il_opt/mem2reg_diamond.in.il
    -DGOLDEN=${_VIPER_GOLDEN_DIR}/il_opt/mem2reg_diamond.opt.il
    -DPASSES=mem2reg
    -P ${_VIPER_GOLDEN_DIR}/il_opt/check_opt.cmake)

  viper_add_ctest(il_opt_mem2reg_nested
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DIL_FILE=${_VIPER_GOLDEN_DIR}/il_opt/mem2reg_nested.in.il
    -DGOLDEN=${_VIPER_GOLDEN_DIR}/il_opt/mem2reg_nested.opt.il
    -DPASSES=mem2reg
    -P ${_VIPER_GOLDEN_DIR}/il_opt/check_opt.cmake)

  viper_add_ctest(il_opt_mem2reg_loop
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DIL_FILE=${_VIPER_GOLDEN_DIR}/il_opt/mem2reg_loop.in.il
    -DGOLDEN=${_VIPER_GOLDEN_DIR}/il_opt/mem2reg_loop.opt.il
    -DPASSES=mem2reg
    -P ${_VIPER_GOLDEN_DIR}/il_opt/check_opt.cmake)

  viper_add_ctest(il_opt_mem2reg_multi
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DIL_FILE=${_VIPER_GOLDEN_DIR}/il_opt/mem2reg_multi.in.il
    -DGOLDEN=${_VIPER_GOLDEN_DIR}/il_opt/mem2reg_multi.opt.il
    -DPASSES=mem2reg
    -P ${_VIPER_GOLDEN_DIR}/il_opt/check_opt.cmake)

  file(READ ${_VIPER_GOLDEN_DIR}/quickstart.stdout EXPECT_QUICKSTART)
  string(STRIP "${EXPECT_QUICKSTART}" EXPECT_QUICKSTART)
  viper_add_ctest(il_quickstart_example
    ${CMAKE_COMMAND}
    -DILC=${BASIC_ILC}
    -DIL_FILE=${_VIPER_GOLDEN_DIR}/quickstart.il
    "-DEXPECT=${EXPECT_QUICKSTART}"
    -P ${_VIPER_GOLDEN_DIR}/check_run_output.cmake)
endfunction()
